src/psyclone/tests/domain/common/extract_driver_creator_test.py

```diff
@@ -0,0 +1,547 @@
+    # It is tested line by line since there is other code in between
+    # which is not important, and the order might also change. It also
+    # tests if unique variable names are created in the driver: the user
+    # program contains a local variable 'dx', which clashes with the grid
+    # property dx. The grid property will be renamed to 'dx_1':
+    expected = '''use extract_psy_data_mod, only : extract_PsyDataType
+
+  real*8, allocatable, dimension(:,:) :: out_fld
+  real*8, allocatable, dimension(:,:) :: in_out_fld
+  real*8, allocatable, dimension(:,:) :: in_fld
+  real*8, allocatable, dimension(:,:) :: dx
+  real*8, allocatable, dimension(:,:) :: in_fld_grid_gphiu
+  real*8, allocatable, dimension(:,:) :: out_fld_post
+  real*8 :: in_fld_grid_dx
+  real*8, allocatable, dimension(:,:) :: in_out_fld_post
+  type(extract_PsyDataType) :: extract_psy_data
+  call extract_psy_data%OpenRead('psy_extract_example_with_various_variable_''' \
+  '''access_patterns', 'invoke_0_compute_kernel:compute_kernel_code:r0')
+  call extract_psy_data%ReadVariable('out_fld_post', out_fld_post)
+  ALLOCATE(out_fld(SIZE(out_fld_post, 1), SIZE(out_fld_post, 2)))
+  out_fld = 0
+  call extract_psy_data%ReadVariable('in_fld', in_fld)
+  call extract_psy_data%ReadVariable('in_out_fld_post', in_out_fld_post)
+  call extract_psy_data%ReadVariable('dx', dx)
+  call extract_psy_data%ReadVariable('in_fld%grid%dx', in_fld_grid_dx)
+  call extract_psy_data%ReadVariable('in_fld%grid%gphiu', in_fld_grid_gphiu)
+  do j = out_fld_internal_ystart, out_fld_internal_ystop, 1
+    do i = out_fld_internal_xstart, out_fld_internal_xstop, 1
+      call compute_kernel_code(i, j, out_fld, in_out_fld, in_fld, dx, ''' \
+      '''in_fld_grid_dx, in_fld_grid_gphiu)
+    enddo
+  enddo
+  if (i == i_post) then
+    PRINT *, "i correct"
+  else
+    PRINT *, "i incorrect. Values are:"
+    PRINT *, i
+    PRINT *, "i values should be:"
+    PRINT *, i_post
+  end if
+  if (ALL(in_out_fld - in_out_fld_post == 0.0)) then
+    PRINT *, "in_out_fld correct"
+  else
+    PRINT *, "in_out_fld incorrect. Values are:"
+    PRINT *, in_out_fld
+    PRINT *, "in_out_fld values should be:"
+    PRINT *, in_out_fld_post
+  end if'''
+    expected_lines = expected.split("\n")
+    for line in expected_lines:
+        assert line in driver_code
+
+
+# -----------------------------------------------------------------------------
+def test_driver_creation2(tmpdir):
+    '''Test different ways of specifying the nodes, including code creation
+    without constant loop boundaries.
+
+    '''
+    # Use tmpdir so that the driver is created in tmp
+    tmpdir.chdir()
+
+    etrans = GOceanExtractTrans()
+    psy, invoke = get_invoke("driver_test.f90",
+                             GOCEAN_API, idx=0, dist_mem=False)
+    schedule = invoke.schedule
+    etrans.apply([schedule.children[0]], {'create_driver': True})
+    # We are only interested in the driver, so ignore results.
+    str(psy.gen)
```

Comment:
OK, I've moved the change of ``create()`` returning a PSyData tree forwards, and am now providing three interfaces: create gets the tree, get_string, and write_driver (which is what is currently happening). I have not replaced all calls, since I need to test that driver creation works for both default and explicitly-specified names. Also, the strange situation is that it's actually the transform that makes sure to avoid duplicated names, but I think this MUST be tested in the driver (i.e. when running the driver creation test I want to check if duplicated names will be avoided, remembering that I would need to run the extract transformation feels wrong. Will discuss this in the telco tonight).